### ISSUE TITLE
Allow custom entries

### ISSUE THREAD

**Issue by CodeSapiens** [aarond10/https_dns_proxy#107]
First of all, thank you much for your hard work and for freely providing a very nice little utility.

Please consider allowing the addition of custom DNS resolvers, or at least allow adding endpoint IDs to the URL.

Your list includes NextDNS.io, which is the one I want to use, but your luci-app-https-dns-proxy does not provide a way for adding a personal endpoint ID, which is what makes NextDNS so valuable and powerful. Without our personal ID the resolver simply runs with basic filters, and we don't benefit from all the custom filters they provide.
A customized NextDNS URL looks like this:  https://dns.nextdns.io/87gh64r

I figured perhaps I could edit the config that contains your list of resolvers, but I stroke out. I searched all over the file structure in my router, and searched your web sources for over an hour, looking for any reference to the location of your config files, and I couldn't find a single reference. Perhaps I overlooked them.

Second related issue...
I was able to add the URL with my my personal endpoint ID via the CLI, based on what I found here: https://openwrt.org/docs/guide-user/services/dns/doh_dnsmasq_https-dns-proxy   and here: https://docs.openwrt.melmac.net/https-dns-proxy

It works as expected, and the Service Status in your luci-app-https-dns-proxy indicates the selected resolver is NextDNS.io DoH at 127.0.0.1:5053, but unfortunately the drop-down menu can't find an exact match, so it just defaults to the first entry and shows "Quad 101 (Taiwan)".

I hope that makes sense. Thanks in advance for your time and effort.

**Comment by moranbw** [comment]
@CodeSapiens I stumbled across NextDNS today and was trying to set it up myself.  I don't see why you can't feed that URL in as your resolver_url (```-r``` per the docs).  I was able to get NextDNS to accept that I was correctly using their service over DoH, but I was getting a lot of failed queries.  So there definitely was something missing on the https_dns_proxy side of things.  I've attached my logs if any of the devs are following this issue.  @aarond10 or @baranyaib90 do you have any thoughts here?

```
* We are completely uploaded and fine
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 51
< date: Sun, 09 May 2021 02:16:38 GMT
< 
* Connection #0 to host dns.nextdns.io left intact
[D] 1620526598.012949 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526598.026140 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526598.026150 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526598.026153 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526598.026155 https_client.c:292 Times: 0.000045, 0.000045, 0.000045, 0.000152, 0.000155, 0.013544
[D] 1620526598.026161 main.c:77 buflen 79

[D] 1620526598.026252 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526598.026255 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526598.026257 https_client.c:292 Times: 0.000000, 0.000000, 0.000000, 0.000111, 0.000114, 0.013329
[D] 1620526598.026261 main.c:77 buflen 91

[D] 1620526598.435076 main.c:94 Received request for id: 9c1e, len: 46
[D] 1620526598.435085 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526598.435353 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526598.454631 https_client.c:401 Released used io event: 0x7ffc084a32e0
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Re-using existing connection! (#0) with host dns.nextdns.io
* Connected to dns.nextdns.io (170.39.224.134) port 443 (#0)
* Using Stream ID: 95 (easy handle 0x7fb0e0c6daf0)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 63

* We are completely uploaded and fine
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Multiplexed connection found!
* Re-using existing connection! (#0) with host dns.nextdns.io
* Using Stream ID: 97 (easy handle 0x7fb0e0d80580)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 63

* We are completely uploaded and fine
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 79
< date: Sun, 09 May 2021 02:16:40 GMT
< 
[D] 1620526598.454641 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526598.454652 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526598.454655 https_client.c:292 Times: 0.000040, 0.000040, 0.000040, 0.000152, 0.000155, 0.019470
[D] 1620526598.454661 main.c:77 buflen 51

[D] 1620526600.065400 main.c:94 Received request for id: 6fd9, len: 63
[D] 1620526600.065409 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526600.065731 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526600.065752 main.c:94 Received request for id: f12b, len: 63
[D] 1620526600.065758 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526600.066036 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526600.066039 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526600.079128 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526600.079132 https_client.c:259 CURLINFO_HTTP_VERSION: 2
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 91
< date: Sun, 09 May 2021 02:16:40 GMT
< 
* Connection #0 to host dns.nextdns.io left intact
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Re-using existing connection! (#0) with host dns.nextdns.io
* Connected to dns.nextdns.io (170.39.224.134) port 443 (#0)
* Using Stream ID: 99 (easy handle 0x7fb0e0d81ab0)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 63

* We are completely uploaded and fine
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Multiplexed connection found!
* Re-using existing connection! (#0) with host dns.nextdns.io
* Using Stream ID: 9b (easy handle 0x7fb0e0c6db00)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 63

* We are completely uploaded and fine
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 79
< date: Sun, 09 May 2021 02:16:42 GMT
< 
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 91
< date: Sun, 09 May 2021 02:16:42 GMT
< 
[D] 1620526600.079135 https_client.c:292 Times: 0.000051, 0.000051, 0.000051, 0.000163, 0.000166, 0.013628
[D] 1620526600.079149 main.c:77 buflen 79

[D] 1620526600.083839 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526600.083848 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526600.083851 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526600.083853 https_client.c:292 Times: 0.000000, 0.000000, 0.000000, 0.000192, 0.000195, 0.018020
[D] 1620526600.083859 main.c:77 buflen 91

[D] 1620526602.131803 main.c:94 Received request for id: abcb, len: 63
[D] 1620526602.131812 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526602.132072 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526602.132093 main.c:94 Received request for id: a658, len: 63
[D] 1620526602.132099 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526602.132247 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526602.132249 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
* Connection #0 to host dns.nextdns.io left intact
[D] 1620526602.146922 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526602.146934 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526602.146937 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526602.146939 https_client.c:292 Times: 0.000041, 0.000041, 0.000042, 0.000147, 0.000151, 0.015011
[D] 1620526602.146946 main.c:77 buflen 79

[D] 1620526602.147017 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526602.147020 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526602.147022 https_client.c:292 Times: 0.000000, 0.000000, 0.000000, 0.000074, 0.000076, 0.014752
[D] 1620526602.147029 main.c:77 buflen 91

* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* 17 bytes stray data read before trying h2 connection
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Re-using existing connection! (#0) with host dns.nextdns.io
* Connected to dns.nextdns.io (170.39.224.134) port 443 (#0)
* Using Stream ID: 9d (easy handle 0x7fb0e0d80590)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 62

* We are completely uploaded and fine
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Multiplexed connection found!
* Re-using existing connection! (#0) with host dns.nextdns.io
* Using Stream ID: 9f (easy handle 0x7fb0e0d81ac0)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 62

* We are completely uploaded and fine
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 78
< date: Sun, 09 May 2021 02:16:44 GMT
< 
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 90
< date: Sun, 09 May 2021 02:16:44 GMT
< 
* Connection #0 to host dns.nextdns.io left intact
[D] 1620526604.197603 main.c:94 Received request for id: 759c, len: 62
[D] 1620526604.197613 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526604.197937 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526604.197962 main.c:94 Received request for id: 28ff, len: 62
[D] 1620526604.197967 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526604.198134 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526604.198136 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526604.213976 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526604.213982 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526604.213985 https_client.c:292 Times: 0.000000, 0.000000, 0.000000, 0.000089, 0.000091, 0.015935
[D] 1620526604.213992 main.c:77 buflen 78

[D] 1620526604.214540 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526604.214548 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Connection 1 seems to be dead!
* Closing connection 1
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Re-using existing connection! (#0) with host dns.nextdns.io
* Connected to dns.nextdns.io (170.39.224.134) port 443 (#0)
* Using Stream ID: a1 (easy handle 0x7fb0e0c6db10)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 62

* We are completely uploaded and fine
* RESOLVE dns.nextdns.io:443 is - old addresses discarded!
* Added dns.nextdns.io:443:170.39.224.134,213.227.173.235 to DNS cache
* Found bundle for host dns.nextdns.io: 0x7fb0e0d96600 [can multiplex]
* Multiplexed connection found!
* Re-using existing connection! (#0) with host dns.nextdns.io
* Using Stream ID: a3 (easy handle 0x7fb0e0d805a0)
> POST /<MY-ID> HTTP/2
Host: dns.nextdns.io
user-agent: dns-to-https-proxy/0.2
accept: application/dns-message
content-type: application/dns-message
content-length: 62

* We are completely uploaded and fine
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 78
< date: Sun, 09 May 2021 02:16:46 GMT
< 
< HTTP/2 200 
< content-type: application/dns-message
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< content-length: 90
< date: Sun, 09 May 2021 02:16:46 GMT
< 
[D] 1620526604.214550 https_client.c:259 CURLINFO_HTTP_VERSION: 2
[D] 1620526604.214556 https_client.c:292 Times: 0.000103, 0.000103, 0.000103, 0.000200, 0.000203, 0.016843
[D] 1620526604.214562 main.c:77 buflen 90

[D] 1620526606.251103 main.c:94 Received request for id: f16f, len: 62
[D] 1620526606.251112 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526606.252227 https_client.c:76 curl closed socket: 9
[D] 1620526606.252361 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526606.252782 main.c:94 Received request for id: 59e6, len: 62
[D] 1620526606.252788 https_client.c:95 Requesting HTTP/1.1: 0

[D] 1620526606.253013 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526606.253016 https_client.c:411 Reserved new io event: 0x7ffc084a32e0
[D] 1620526606.271000 https_client.c:401 Released used io event: 0x7ffc084a32e0
[D] 1620526606.271010 https_client.c:245 CURLINFO_EFFECTIVE_URL: https://dns.nextdns.io/<MY-ID>
[D] 1620526606.271013 https_client.c:259 CURLINFO_HTTP_VERSION: 2
```

**Comment by baranyaib90** [comment]
Hi,
I have checked the logs and found nothing special.
No connection/HTTP/other errors at all :S
Could you please do the following:
1. Try to tcpdump the DNS traffic to the HTTPS proxy. It may happen, that NextDNS replies with DNS resolve failures. (That wouldn't be the fault of the HTTPS proxy.)
2. Try to pinpoint, which hostnames fail to resolve and if the resolve failures are constant or random.
3. Update the code to latest master, since there were added HTTP request faulire debugging improvements.

**Comment by moranbw** [comment]
@baranyaib90 thanks for having a look.  After some more digging and tweaking, it turned out it was an issue with my DNSSEC config in dnsmasq (which sits between my network and https_dns_proxy).

So in my view, NextDNS works just fine with the custom endpoints over DoH.  You just need to provide your custom URL as resolver_url, and the IP addresses that are mentioned in the "Setup Guide" (i.e. not the "Linked IP" addresses).

tldr I think this issue can be closed.  @CodeSapiens seems to be referring to the OpenWRT package, which seems to only make certain resolvers available.  It'd be on whoever maintains that package to make it more configurable, right?

**Comment by aarond10** [comment]
Yep! @stangri is the person in you're after for the OpenWRT config.
